### PR TITLE
[MRG+1] Support parallel fitting to GradientBoostingQuantileRegressor

### LINKED ISSUE #141: n_jobs parameter to GradientBoostingQuantileRegressor
Fitting across the three quantiles (or provided) should be embarassingly parallel.

### PR DESCRIPTION
**Pull request by nel215**
I attempted to fix #141.

### CODE REVIEW

**Comment by betatim** [comment]
Thanks a lot for the PR. I will take a look at it now.

@MechCoder @glouppe should we have our own `externals` or just use the `sklearn` one or ...? Not sure what is the normal thing to do.

**Comment by MechCoder** [comment]
I think for joblib it is all right to use `sklearn.externals`

File: skopt/learning/tests/test_gbrt.py [hunk 1]
```diff
@@ -94,3 +94,22 @@ def test_gbrt_with_std():
...
+
+    rgr = GradientBoostingQuantileRegressor(n_jobs=1)
+    rgr.fit(X, y)
+    estimates = rgr.predict(X)
+
+    rgr.set_params(n_jobs=2)
+    rgr.fit(X, y)
+    estimates_parallel = rgr.predict(X)
+
+    assert_almost_equal(np.mean(estimates, axis=0),
```
Comment (MechCoder) [inline skopt/learning/tests/test_gbrt.py]: You can check `estimates` directly right?

**Comment by MechCoder** [comment]
looks ok to me

File: skopt/learning/gbrt.py [hunk 2]
```diff
@@ -27,13 +32,18 @@ class GradientBoostingQuantileRegressor(BaseEstimator, RegressorMixin):
...
         results.
+
+    * `n_jobs` [int, default=1]:
+        The number of jobs to run in parallel for `fit`.
+        If -1, then the number of jobs is set to the number of cores.
     """
 
     def __init__(self, quantiles=[0.16, 0.5, 0.84], base_estimator=None,
-                 random_state=None):
+                 random_state=None, n_jobs=1):
```
Comment (betatim) [inline skopt/learning/gbrt.py]: Could you swap the order of `random_state` and `n_jobs`?

File: skopt/learning/tests/test_gbrt.py [hunk 3]
```diff
@@ -94,3 +94,22 @@ def test_gbrt_with_std():
...
+                        np.mean(estimates_parallel, axis=0))
```
Comment (betatim) [inline skopt/learning/tests/test_gbrt.py]: If you set (and reset) `random_state` for the regressor the return value from `fit()` should be exactly the same, which you can compare with `assert_almost_equal` instead comparing the summary statistic (the mean).

**Comment by betatim** [comment]
I think we can use the [`threading`](https://pythonhosted.org/joblib/parallel.html#using-the-threading-backend) backend with GBT which is more lightweight than `multiprocessing`.

Can you add a parameter to `gbrt_minimize` (and `forest_minimize` for ET and RF) that allows the user to control the number of jobs?

**Comment by nel215** [comment]
Thanks for the comments! I'll fix these.

**Comment by nel215** [comment]
I made some fixes.

Due to the change of handling random_state, the `test_tree_based_minimize` test failed.
So, I increased `n_call` of the test. Do you have a better idea than this?

**Comment by betatim** [comment]
Why do you think we need [different seeds](https://github.com/scikit-optimize/scikit-optimize/pull/165/files#diff-011506ca7d0237d223cbe785bf810be7R80) for each of the regressors?

The thinking behind using the same was to try and ensure that the quantiles will always be sorted. The predictions for different quantiles (here the Xth quantile and the X+eps'th quantile) should be sorted: `q(X) < quantile(X+epsilon)` which if the two regressors are "independent" (because they have different seeds) might not be true. I never benchmarked this to see how big a problem this is in practice.

**Comment by nel215** [comment]
I may misunderstand it. So let me ask you some questions.

If it is necessary to use the same seed for each of the regressors, the master branch wrong?
Should the following code output the same integer values?

```
> cat a.py
from sklearn.utils import check_random_state
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.base import clone
import numpy as np

random_state = np.random.RandomState(1)
rng = check_random_state(random_state)

base_estimator = GradientBoostingRegressor(loss='quantile')
quantiles=[0.16, 0.5, 0.84]

for q in quantiles:
    regressor = clone(base_estimator)
    regressor.set_params(alpha=q, random_state=rng)
    print(regressor.random_state.randint(10))
> python a.py
5
8
9
```

If it is necessary to output same integer values, I think that the following code is required.

```
> cat b.py
from sklearn.utils import check_random_state
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.base import clone
import numpy as np

random_state = np.random.RandomState(1)
rng = check_random_state(random_state)

base_estimator = GradientBoostingRegressor(loss='quantile')
quantiles=[0.16, 0.5, 0.84]

base_estimator.set_params(random_state=rng)
for q in quantiles:
    regressor = clone(base_estimator)
    regressor.set_params(alpha=q)
    print(regressor.random_state.randint(10))
> python a.py
5
5
5
```

Is this correct?

**Comment by betatim** [comment]
Yes indeed `master` is wrong in this case! Good catch. My intention was definitely to have `regressor.random_state.randint(10)` print the same number for each regressor.

When you change the code to do that, can you add a comment that summarizes our discussion so that in the future people reading the code know why we want each regressor to produce the same random numbers.

File: skopt/tests/test_tree_opt.py [hunk 4]
```diff
@@ -91,6 +91,6 @@ def test_tree_based_minimize():
         yield (check_minimize, minimizer, bench4, 0.0,
                [("-2", "-1", "0", "1", "2")], 0.05, 10)
         yield (check_minimize, minimizer, branin, 0.39,
-               [(-5.0, 10.0), (0.0, 15.0)], 0.1, 100)
+               [(-5.0, 10.0), (0.0, 15.0)], 0.1, 160)
```
Comment (glouppe) [inline skopt/tests/test_tree_opt.py]: I did some changes recently on this, you will need to rebase

**Comment by nel215** [comment]
Ok. I'll fix it.

**Comment by nel215** [comment]
I fixed.

**Comment by glouppe** [comment]
Thanks a lot for your contribution @nel215, this looks good to me! What do you think @betatim ?

**Comment by betatim** [comment]
Merged! Thanks for the contribution and thanks for spotting&squashing that bug.

I added you to [`AUTHORS.md`](https://github.com/scikit-optimize/scikit-optimize/blob/master/AUTHORS.md) and ['CHANGELOG.md`](https://github.com/scikit-optimize/scikit-optimize/blob/master/CHANGELOG.md)

**Comment by nel215** [comment]
Thanks for the review :)
